Various fixes for timing counts

pchintalapudi opened the pull request:
This bundles a number of fixes and cleanups for using the timing counts profiling support.

KristofferC commented:
I'll defer the review here to @topolarity who is more familiar with the tracy stuff and its current implementation in Julia.

topolarity commented:
Self and total times is very nice. Can you share an example output?

topolarity commented:
> Here's an example output (with some addition from mimicking https://github.com/JuliaLang/julia/pull/49675): https://github.com/JuliaLang/julia/pull/49675#issuecomment-1539204623

Hmm I'm a bit hesitant to add so many events to the COUNTS output, but I rarely use it so my opinion doesn't matter too much

Maybe @vtjnash or @vchuravy have an opinion here

pchintalapudi committed:
Make jl_timing_counts atomic

pchintalapudi committed:
Print timing outputs as CSV

pchintalapudi committed:
Make jl_timing_enable_mask atomic

pchintalapudi committed:
jl_timing_enable_mask -> jl_timing_disable_mask

pchintalapudi committed:
Allow number of subsystems to exceed 64

pchintalapudi committed:
Sort timing names

pchintalapudi committed:
Print total time and self timing with timing counts

pchintalapudi committed:
Overwrite random value in root timing slot

vchuravy commented:
> Hmm I'm a bit hesitant to add so many events to the COUNTS output, but I rarely use it so my opinion doesn't matter too much

I don't have a strong opinion. I tend to see @JeffBezanson using them :)

topolarity commented:
IIUC the latest version aggregates counts by subsystem again so the only change to the counts output is that it reports self and total time now?

The `jl_timing_disable_mask` expansion makes my YAGNI reflex itch a bit (I'm not sure we'll want more than 64 subsystems in the long-term and it feels like the change gave up some readability), but the other improvements are quite nice, including using proper atomics etc.


Also, somewhat orthogonal idea but what do you think about trying to report these "counts" in units of time, since they are not actual core cycles? (they are TSC cycles)

pchintalapudi commented:
Oh right I should change that to events, we won't lose any granularity since total time exists now. I would very much like to report proper times rather than cycles, that seems like a more useful metric.

topolarity commented:
> I would very much like to report proper times rather than cycles

For reference, this is how Tracy measures the TSC frequency: https://github.com/wolfpld/tracy/blob/f3d3108d5fe882ec93926808ecee2f7f6c761f3b/public/client/TracyProfiler.cpp#LL3524C1-L3540C6

Although the TSC frequency  is fixed and known by the OS kernel, unfortunately I don't know of any good way to get it without measuring it yourself

pchintalapudi committed:
Time events instead of subsystems

pchintalapudi commented:
I'll leave the cycle->time change to a separate PR, but I was thinking we could just swap the cycleclock() calls to jl_hrtime() calls. It may be slightly more expensive to do compared to reading a register, but I think the interpretability benefits will outweigh the performance loss.

topolarity commented on src/timing.c:
```diff
@@ -325,20 +350,27 @@ void jl_timing_init_task(jl_task_t *t)
+    int i = *idx;
+    // sorted names include events, so skip if we're looking at an event instead of a subsystem
+    // events are always at least JL_TIMING_LAST
+    if (i >= JL_TIMING_LAST)
```
Hmm, what guarantees that the subsystems are the first `JL_TIMING_LAST` items after sorting?

pchintalapudi replied:
This is guaranteed by the JL_TIMING_EVENTS macro, which puts subsystems before events.

topolarity replied:
Isn't that affected by the sort?

topolarity replied:
Oh I just saw that `jl_timing_names_sorted` array is an array of indices, not a sorted version of the original array

Understanding that, this looks correct to me.

topolarity commented:
> I'll leave the cycle->time change to a separate PR

Sounds good to me 👍 

>  I was thinking we could just swap the cycleclock() calls to jl_hrtime() calls. It may be slightly more expensive to do compared to reading a register, but I think the interpretability benefits will outweigh the performance loss.

I think pretty much any timer that provides nanosecond resolution on x86 is implemented on top of the TSC. For example, macOS provides `mach_absolute_time` which is what jl_hrtime calls through libuv's uv_hrtime on macOS. That [function](https://github.com/apple/darwin-xnu/blob/2ff845c2e033bd0ff64b5b6aa6063a1f8f65aa32/osfmk/i386/rtclock.c#L483-L486) is [implemented with `rdtsc`](https://github.com/apple/darwin-xnu/blob/2ff845c2e033bd0ff64b5b6aa6063a1f8f65aa32/osfmk/i386/rtclock_asm_native.h#L42-L82). macOS just happens to be one of the platforms nice enough to also provide a way for us to get the frequency with mach_timebase_info.

Not that that means we can't use libuv's `uv_hrtime` - Just wanted to fill in some details of the timer stack

pchintalapudi commented on src/timing.h:
```diff
@@ -300,33 +302,35 @@ STATIC_INLINE void _jl_timing_counts_stop(jl_timing_counts_t *block, uint64_t t)
-STATIC_INLINE void _jl_timing_counts_ctor(jl_timing_counts_t *block) JL_NOTSAFEPOINT {
+STATIC_INLINE void _jl_timing_counts_ctor(jl_timing_counts_t *block, uint64_t t) JL_NOTSAFEPOINT {
     block->total = 0;
+    block->t0 = t;
```
It needs to be in the constructor because we start and stop a block whenever we have a nested block, but we only construct it once. We shouldn't be relying on the time difference between construction and starting or else we risk not accounting for some time when profiling.

topolarity replied:
I mean to put this in `jl_timing_block_start` which is only called once per block

pchintalapudi committed:
Move t0 init to jl_timing_block_start

pchintalapudi committed:
Update src/timing.h

Co-authored-by: Cody Tapscott <84105208+topolarity@users.noreply.github.com>

pchintalapudi committed:
t0 is in counts_ctx

topolarity commented on src/timing.h:
```diff
@@ -351,6 +353,9 @@ STATIC_INLINE void jl_timing_block_start(jl_timing_block_t *block) {
     uint64_t t = cycleclock(); (void)t;
     _COUNTS_START(&block->counts_ctx, t);
+#ifdef USE_TIMING_COUNTS
+    block->counts_ctx.t0 = t;
```
Why not put this in `_COUNTS_START`?

topolarity replied:
Ah of course, you don't want this to run in `_jl_timing_block_destroy`

topolarity replied:
I would be open to this re-factoring:
- `_COUNTS_START   => _COUNTS_START (only called once) + _COUNTS_RESUME (resumes self timer)`
- `_COUNTS_STOP    => _COUNTS_PAUSE (stops self timer)`
- `_COUNTS_DESTROY => _COUNTS_END (stops timer and adds counts to totals)`

Hopefully that also makes it clearer that _COUNTS_END should be called conditionally if the block was running, not unconditionally to destroy the object itself

topolarity commented on src/timing.h:
```diff
@@ -318,16 +320,16 @@ STATIC_INLINE void _jl_timing_counts_ctor(jl_timing_counts_t *block) JL_NOTSAFEP
-    jl_timing_counts[subsystem] += block->total;
+STATIC_INLINE void _jl_timing_counts_destroy(jl_timing_counts_t *block, int event, uint64_t t) JL_NOTSAFEPOINT {
+    jl_atomic_fetch_add_relaxed(jl_timing_self_counts + event, block->total);
+    jl_atomic_fetch_add_relaxed(jl_timing_full_counts + event, t - block->t0);
```
`block->t0` might be uninitialized if the event was created but not started, but this function is supposed to do nothing in that case

The smoking gun I noticed was this:
```
GC_IncrementalSweep      ,             24332148 ( 0.23 %),      158290379112508 (1464937.27 %)
```

topolarity committed:
Slightly re-factor timing counts back-end

This should make it clearer what the responsibilities for each function
are, and fix a bug where conditionally-triggered events could end up with
an invalid total time.

topolarity commented:
Example output:
```
JULIA TIMINGS
Event                    , Self Cycles (% of Total)      , Total Cycles (% of Total)
ADD_METHOD               ,            157925151 ( 0.05 %),            157939409 ( 0.05 %)
CODEGEN_Codeinst         ,            189261109 ( 0.06 %),           5441888779 ( 1.60 %)
CODEGEN_LLVM             ,           5254311020 ( 1.55 %),           5254956080 ( 1.55 %)
CODEGEN_Workqueue        ,             16346963 ( 0.00 %),           1143138561 ( 0.34 %)
CODEINST_COMPILE         ,            823215342 ( 0.24 %),          62866917537 (18.52 %)
DL_OPEN                  ,                29640 ( 0.00 %),                29640 ( 0.00 %)
GC                       ,             28728718 ( 0.01 %),           4425692594 ( 1.30 %)
GC_Finalizers            ,               427748 ( 0.00 %),               427748 ( 0.00 %)
GC_FullSweep             ,            119581722 ( 0.04 %),            119581722 ( 0.04 %)
GC_IncrementalSweep      ,           3045079865 ( 0.90 %),           3045079865 ( 0.90 %)
GC_Mark                  ,           1232335535 ( 0.36 %),           1232335535 ( 0.36 %)
GC_Stop                  ,               372552 ( 0.00 %),               372552 ( 0.00 %)
INFERENCE                ,           2474823100 ( 0.73 %),          13859252913 ( 4.08 %)
JULIA_INIT               ,             44863046 ( 0.01 %),            515658464 ( 0.15 %)
LLVM_OPT                 ,           8162171554 ( 2.40 %),           8162171554 ( 2.40 %)
LLVM_ORC                 ,          48441573182 (14.27 %),          56601762669 (16.67 %)
LOWERING                 ,          10244842130 ( 3.02 %),          12334152525 ( 3.63 %)
MACRO_INVOCATION         ,            240035794 ( 0.07 %),           2071023077 ( 0.61 %)
METHOD_LOOKUP_SLOW       ,            180046169 ( 0.05 %),            180596227 ( 0.05 %)
PARSING                  ,           4557581897 ( 1.34 %),           4557581897 ( 1.34 %)
ROOT                     ,         253859209642 (74.77 %),         339517920072 (100.00 %)
STAGED_FUNCTION          ,              3827616 ( 0.00 %),             48451920 ( 0.01 %)
SYSIMG_DUMP              ,            425928008 ( 0.13 %),            425928008 ( 0.13 %)
TYPE_CACHE_INSERT        ,             15402569 ( 0.00 %),             15402569 ( 0.00 %)

JULIA COUNTERS
Counter                  , Value
HeapSize                 ,             19503084
JITSize                  ,              8584384
JITCodeSize              ,              7672825
JITDataSize              ,               911559
```

topolarity reviewed (approved):
Ready to merge once CI passes
